library/psa_crypto.c

```diff
@@ -3131,7 +3135,690 @@ psa_status_t psa_asymmetric_decrypt( mbedtls_svc_key_id_t key,
+                            mbedtls_svc_key_id_t key, psa_algorithm_t alg,
+                            const uint8_t *hash, size_t hash_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot;
+
+    /* Check that start has not been previously called. */
+    if( operation->id != 0 )
+        return( PSA_ERROR_BAD_STATE );
+
+
+    status = psa_sign_verify_check_alg( 0, alg );
+    if( status != PSA_SUCCESS )
+        return status;
+
+    status = psa_get_and_lock_key_slot_with_policy( key, &slot,
+                                                    PSA_KEY_USAGE_SIGN_HASH,
+                                                    alg );
+
+    if( status != PSA_SUCCESS )
+        goto exit;
+
+    if( !PSA_KEY_TYPE_IS_KEY_PAIR( slot->attr.type ) )
+    {
+        status = PSA_ERROR_INVALID_ARGUMENT;
+        goto exit;
+    }
+
+    psa_key_attributes_t attributes = {
+        .core = slot->attr
+    };
+
+    /* Ensure ops count gets reset, in case of operation re-use. */
+    operation->num_ops = 0;
+
+    status = psa_driver_wrapper_sign_hash_start( operation, &attributes,
+                                                 slot->key.data,
+                                                 slot->key.bytes, alg,
+                                                 hash, hash_length );
+exit:
+
+    if( status != PSA_SUCCESS )
+        psa_sign_hash_abort_internal( operation );
+
+    unlock_status = psa_unlock_key_slot( slot );
+
+    return( ( status == PSA_SUCCESS ) ? unlock_status : status );
+
+}
+
+
+psa_status_t psa_sign_hash_complete(
+                            psa_sign_hash_interruptible_operation_t *operation,
+                            uint8_t *signature, size_t signature_size,
+                            size_t *signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    *signature_length = 0;
+
+    /* Check that start has been called first. */
+    if( operation->id == 0 )
+    {
+        status = PSA_ERROR_BAD_STATE;
+        goto exit;
+    }
+
+    /* Immediately reject a zero-length signature buffer. This guarantees
+    * that signature must be a valid pointer. (On the other hand, the input
+    * buffer can in principle be empty since it doesn't actually have
+    * to be a hash.) */
+    if( signature_size == 0 )
+    {
+        status = PSA_ERROR_BUFFER_TOO_SMALL;
+        goto exit;
+    }
+
+    status = psa_driver_wrapper_sign_hash_complete( operation, signature,
+                                                    signature_size,
+                                                    signature_length );
+exit:
+
+    /* Update ops count with work done. */
+    operation->num_ops += psa_driver_wrapper_sign_hash_get_num_ops( operation );
+
+    if( status != PSA_OPERATION_INCOMPLETE )
+    {
+        /* Fill the unused part of the output buffer (the whole buffer on error,
```

Comment:
If you mean the code in `psa_sign_internal()`, its not exactly the same as that, the else condition is different for starters due to differing constraints. The only bit that can really be made into an auxiliary function is the `memset()`, which seems not worth it.

Reply:
Yes, I mean `psa_sign_internal`. They can share a `wipe_output` function. (This may be preexisting in that hash and MAC can probably use the same code.)